proxy/src/main/java/com/wavefront/agent/ProxyCheckinScheduler.java

```diff
@@ -0,0 +1,264 @@
+   */
+  public void shutdown() {
+    executor.shutdown();
+  }
+
+  /**
+   * Perform agent check-in and fetch configuration of the daemon from remote server.
+   *
+   * @return Fetched configuration. {@code null} if the configuration is invalid.
+   */
+  private AgentConfiguration checkin() {
+    AgentConfiguration newConfig;
+    JsonNode agentMetricsWorkingCopy;
+    synchronized (executor) {
+      if (agentMetrics == null) return null;
+      agentMetricsWorkingCopy = agentMetrics;
+      agentMetrics = null;
+      if (retries.incrementAndGet() > MAX_CHECKIN_ATTEMPTS) return null;
+    }
+    logger.info("Checking in: " + ObjectUtils.firstNonNull(serverEndpointUrl,
+        proxyConfig.getServer()));
+    try {
+      newConfig = apiContainer.getProxyV2API().proxyCheckin(proxyId,
+          "Bearer " + proxyConfig.getToken(), proxyConfig.getHostname(), getBuildVersion(),
+          System.currentTimeMillis(), agentMetricsWorkingCopy, proxyConfig.isEphemeral());
+      agentMetricsWorkingCopy = null;
+    } catch (ClientErrorException ex) {
+      agentMetricsWorkingCopy = null;
+      switch (ex.getResponse().getStatus()) {
+        case 401:
+          checkinError("HTTP 401 Unauthorized: Please verify that your server and token settings" +
+              " are correct and that the token has Proxy Management permission!");
+          break;
+        case 403:
+          checkinError("HTTP 403 Forbidden: Please verify that your token has Proxy Management " +
+              "permission!");
+          break;
+        case 404:
+        case 405:
+          String serverUrl = proxyConfig.getServer().replaceAll("/$", "");
+          if (succesfulCheckins.get() == 0 && !retryImmediately && !serverUrl.endsWith("/api")) {
+            this.serverEndpointUrl = serverUrl + "/api/";
+            checkinError("Possible server endpoint misconfiguration detected, attempting to use " +
+                serverEndpointUrl);
+            apiContainer.updateServerEndpointURL(serverEndpointUrl);
+            retryImmediately = true;
+            return null;
+          }
+          String secondaryMessage = serverUrl.endsWith("/api") ?
+              "Current setting: " + proxyConfig.getServer() :
+              "Server endpoint URLs normally end with '/api/'. Current setting: " +
+                  proxyConfig.getServer();
+          checkinError("HTTP " + ex.getResponse().getStatus() + ": Misconfiguration detected, " +
+              "please verify that your server setting is correct. " + secondaryMessage);
+          if (succesfulCheckins.get() == 0) {
+            logger.warning("Aborting start-up");
+            System.exit(-5);
+          }
+          break;
+        case 407:
+          checkinError("HTTP 407 Proxy Authentication Required: Please verify that " +
+              "proxyUser and proxyPassword settings are correct and make sure your HTTP proxy" +
+              " is not rate limiting!");
+          break;
```

Comment:
Candidate for 407 since changing proxyUser and proxyPassword requires restart anyhow (like 405 which is likely a misconfiguration)